iocell/src/main/scala/chisel/IOCell.scala

```diff
@@ -0,0 +1,188 @@
+          val iocell = anaFn()
+          name.foreach(n => iocell.suggestName(n))
+          iocell.io.core <> coreSignal
+          padSignal <> iocell.io.pad
+          Seq(iocell)
+        }
+      }
+      case coreSignal: Clock => {
+        DataMirror.directionOf(coreSignal) match {
+          case ActualDirection.Input => {
+            val iocell = inFn()
+            name.foreach(n => iocell.suggestName(n))
+            coreSignal := iocell.io.i.asClock
+            iocell.io.ie := true.B
+            iocell.io.pad := padSignal.asUInt.asBool
+            Seq(iocell)
+          }
+          case ActualDirection.Output => {
+            val iocell = outFn()
+            name.foreach(n => iocell.suggestName(n))
+            iocell.io.o := coreSignal.asUInt.asBool
+            iocell.io.oe := true.B
+            padSignal := iocell.io.pad.asClock
+            Seq(iocell)
+          }
+          case _ => throw new Exception("Unknown direction")
+        }
+      }
+      case coreSignal: Bits => {
+        require(padSignal.getWidth == coreSignal.getWidth, "padSignal and coreSignal must be the same width")
+        if (padSignal.getWidth == 0) {
+          // This dummy assignment will prevent invalid firrtl from being emitted
+          DataMirror.directionOf(coreSignal) match {
+            case ActualDirection.Input => coreSignal := 0.U
+            case _ => {}
+          }
+          Seq()
+        } else {
+          DataMirror.directionOf(coreSignal) match {
+            case ActualDirection.Input => {
+              // this type cast is safe because we guarantee that padSignal and coreSignal are the same type (T), but the compiler is not smart enough to know that
+              val iocells = padSignal.asInstanceOf[Bits].asBools.zipWithIndex.map { case (w, i) =>
+                val iocell = inFn()
+                name.foreach(n => iocell.suggestName(n + "_" + i))
+                iocell.io.pad := w
+                iocell.io.ie := true.B
+                iocell
+              }
+              coreSignal := Cat(iocells.map(_.io.i).reverse)
+              iocells
+            }
+            case ActualDirection.Output => {
+              val iocells = coreSignal.asBools.zipWithIndex.map { case (w, i) =>
+                val iocell = outFn()
+                name.foreach(n => iocell.suggestName(n + "_" + i))
+                iocell.io.o := w
+                iocell.io.oe := true.B
+                iocell
+              }
+              padSignal := Cat(iocells.map(_.io.pad).reverse)
+              iocells
+            }
+            case _ => throw new Exception("Unknown direction")
+          }
+        }
+      }
+      case coreSignal: Vec[Data] => {
+        // this type cast is safe because we guarantee that padSignal and coreSignal are the same type (T), but the compiler is not smart enough to know that
```

Comment:
I wonder if there is a way to match on both in the case to avoid the cast?

Reply:
I can try matching on a tuple of both. Let me play with that.

Reply:
Cool, the tuple worked.